line_profiler/autoprofile/autoprofile.py

```diff
@@ -92,10 +97,43 @@ def run(script_file, ns, prof_mod, profile_imports=False, as_module=False):
         as_module (bool):
             Whether we're running script_file as a module
     """
-    Profiler = AstTreeModuleProfiler if as_module else AstTreeProfiler
+    @contextlib.contextmanager
+    def restore_dict(d, target=None):
+        copy = d.copy()
+        yield target
+        d.clear()
+        d.update(copy)
+
+    if as_module:
+        Profiler = AstTreeModuleProfiler
+        module_name = modpath_to_modname(script_file)
+        assert module_name is not None
+
+        module_obj = types.ModuleType(module_name)
+        namespace = vars(module_obj)
+        namespace.update(ns)
+
+        # Set the `__spec__` correctly
+        spec = getattr(sys.modules.get(module_name), '__spec__', None)
+        if spec is None:
+            spec = importlib.util.find_spec(module_name)
+        module_obj.__spec__ = spec
+
+        # Set the module object to `sys.modules` via a callback, and
+        # then restore it via the context manager
+        callback = functools.partial(
```

Comment:
Similar style thing here. I probably would have implemented this more explicitly in a class-based context manager. But I also think it's clear enough as-is.